spark-submission-worker/src/main/java/org/apache/spark/k8s/operator/SparkAppSubmissionWorker.java

```diff
@@ -0,0 +1,147 @@
+      }
+    }
+    effectiveSparkConf.set("spark.kubernetes.namespace", app.getMetadata().getNamespace());
+    MainAppResource primaryResource = new JavaMainAppResource(Option.empty());
+    if (StringUtils.isNotEmpty(applicationSpec.getJars())) {
+      primaryResource = new JavaMainAppResource(Option.apply(applicationSpec.getJars()));
+      effectiveSparkConf.setIfMissing("spark.jars", applicationSpec.getJars());
+    } else if (StringUtils.isNotEmpty(applicationSpec.getPyFiles())) {
+      primaryResource = new PythonMainAppResource(applicationSpec.getPyFiles());
+      effectiveSparkConf.setIfMissing("spark.submit.pyFiles", applicationSpec.getPyFiles());
+    } else if (StringUtils.isNotEmpty(applicationSpec.getSparkRFiles())) {
+      primaryResource = new RMainAppResource(applicationSpec.getSparkRFiles());
+    }
+    effectiveSparkConf.setIfMissing(
+        "spark.master", "k8s://https://$KUBERNETES_SERVICE_HOST:$KUBERNETES_SERVICE_PORT");
+    String appId = generateSparkAppId(app);
+    effectiveSparkConf.setIfMissing("spark.app.id", appId);
+    return SparkAppDriverConf.create(
+        effectiveSparkConf,
+        appId,
+        primaryResource,
+        applicationSpec.getMainClass(),
+        applicationSpec.getDriverArgs().toArray(new String[0]),
+        Option.apply(applicationSpec.getProxyUser()));
+  }
+
+  protected SparkAppResourceSpec buildResourceSpec(
+      SparkAppDriverConf kubernetesDriverConf, KubernetesClient client) {
+    KubernetesDriverBuilder builder = new KubernetesDriverBuilder();
+    KubernetesDriverSpec kubernetesDriverSpec =
+        builder.buildFromFeatures(kubernetesDriverConf, client);
+    return new SparkAppResourceSpec(kubernetesDriverConf, kubernetesDriverSpec);
+  }
+
+  /**
+   * Generate a Spark application id. Similar to `KubernetesConf.getKubernetesAppId()`. This is
+   * deterministic per attempt per Spark App in order to ensure operator reconciliation idempotency
+   */
+  public static String generateSparkAppId(final SparkApplication app) {
+    long attemptId = ModelUtils.getAttemptId(app);
+    String preferredId = String.format("%s-%d", app.getMetadata().getName(), attemptId);
+    if (preferredId.length() > DEFAULT_ID_LENGTH_LIMIT) {
+      int preferredIdPrefixLength =
+          DEFAULT_ID_LENGTH_LIMIT - DEFAULT_HASH_BASED_IDENTIFIER_LENGTH_LIMIT - 1;
+      String preferredIdPrefix = preferredId.substring(0, preferredIdPrefixLength);
+      return generateHashBasedId(
+          preferredIdPrefix,
+          app.getMetadata().getNamespace(),
+          app.getMetadata().getName(),
+          String.valueOf(attemptId));
+    } else {
+      return preferredId;
+    }
+  }
+
+  public static String generateHashBasedId(final String prefix, final String... identifiers) {
+    return generateHashBasedId(
+        prefix, DEFAULT_ENCODE_BASE, DEFAULT_HASH_BASED_IDENTIFIER_LENGTH_LIMIT, identifiers);
+  }
+
+  public static String generateHashBasedId(
```

Comment:
This method looks a little unsafe because of varargs, `identifiers`.

- Could you provide a way to cap the maximum string length always?
- Also, please add a method description because this is a `public static` method.